src/preprocessing/passes/bv_to_int.cpp

```diff
@@ -0,0 +1,981 @@
+              Node plus = d_nm->mkNode(kind::PLUS, intized_children);
+              Node multSig = d_nm->mkNode(kind::MULT, sigma, pow2(bvsize));
+              d_bvToIntCache[current] =
+                  d_nm->mkNode(kind::MINUS, plus, multSig);
+              d_rangeAssertions.insert(d_nm->mkNode(kind::LEQ, d_zero, sigma));
+              d_rangeAssertions.insert(d_nm->mkNode(kind::LEQ, sigma, d_one));
+              d_rangeAssertions.insert(
+                  mkRangeConstraint(d_bvToIntCache[current], bvsize));
+              break;
+            }
+            case kind::BITVECTOR_MULT:
+            {
+              uint64_t bvsize = current[0].getType().getBitVectorSize();
+              /**
+               * we use a similar trick to the one used for addition.
+               * Tr(a*b) is Tr(a)*Tr(b)-(sigma*2^k),
+               * with k being the bit width,
+               * and sigma is between [0, 2^k - 1).
+               */
+              Node sigma = d_nm->mkSkolem(
+                  "__bvToInt_sigma_var",
+                  d_nm->integerType(),
+                  "Variable introduced in bvToInt pass to avoid integer mod");
+              Node mult = d_nm->mkNode(kind::MULT, intized_children);
+              Node multSig = d_nm->mkNode(kind::MULT, sigma, pow2(bvsize));
+              d_bvToIntCache[current] =
+                  d_nm->mkNode(kind::MINUS, mult, multSig);
+              d_rangeAssertions.insert(
+                  mkRangeConstraint(d_bvToIntCache[current], bvsize));
+              d_rangeAssertions.insert(mkRangeConstraint(sigma, bvsize));
+              break;
+            }
+            case kind::BITVECTOR_UDIV_TOTAL:
+            {
+              uint64_t bvsize = current[0].getType().getBitVectorSize();
+              // we use an ITE for the case where the second operand is 0.
+              Node pow2BvSize = pow2(bvsize);
+              Node divNode =
+                  d_nm->mkNode(kind::INTS_DIVISION_TOTAL, intized_children);
+              Node ite = d_nm->mkNode(
+                  kind::ITE,
+                  d_nm->mkNode(kind::EQUAL, intized_children[1], d_zero),
+                  d_nm->mkNode(kind::MINUS, pow2BvSize, d_one),
+                  divNode);
+              d_bvToIntCache[current] = ite;
+              break;
+            }
+            case kind::BITVECTOR_UREM_TOTAL:
+            {
+              // we use an ITE for the case where the second operand is 0.
+              Node modNode =
+                  d_nm->mkNode(kind::INTS_MODULUS_TOTAL, intized_children);
+              Node ite = d_nm->mkNode(
+                  kind::ITE,
+                  d_nm->mkNode(kind::EQUAL, intized_children[1], d_zero),
+                  intized_children[0],
+                  modNode);
+              d_bvToIntCache[current] = ite;
+              break;
+            }
+            case kind::BITVECTOR_NEG:
+            {
+              // we use an ITE for the case where the operand is 0.
```

Comment:
Why is this case split necessary here?

Reply:
This is an optimization to avoid introducing a `mod`. I now document it in the code.